### ISSUE TITLE
Design Considerations for the texture_index_mapper Function

### ISSUE THREAD

**Issue by GinmusWJQ** [splashdust/bevy_voxel_world#38]
Dear bevy_voxel_world Development Team,
First and foremost, thank you for developing this excellent library, which brings great value to the Bevy ecosystem. While using it, I noticed an interesting design detail in the texture_index_mapper function that I'd like to discuss with you.
The current implementation is as follows:
```rust
fn texture_index_mapper(&self) -> Arc<dyn Fn(u8) -> [u32; 3] + Send + Sync> {
    Arc::new(|mat| match mat {
        0 => [0, 0, 0],
        1 => [1, 1, 1],
        2 => [2, 2, 2],
        3 => [3, 3, 3],
        _ => [0, 0, 0],
    })
}
```
I noticed that the input parameter `mat` is of type `u8`, while the returned texture indices are of type `u32`. This design seems to limit the number of available material types to 256, while the texture indices can reach the maximum value of `u32`.
I have a few questions I'd like to ask:

1. What considerations led to this design? Are there specific reasons related to performance or memory usage?
2. For projects requiring more than 256 material types, are there any recommended best practices?
3. Has there been any consideration to allow users to customize the input parameter type, perhaps through generics or traits?
4. If this is indeed a limitation, are there any plans to adjust it in future versions?

I understand that this may involve underlying design decisions and could potentially impact performance or other aspects. My purpose in raising these questions is to better understand the library's design philosophy and to explore whether it's possible to provide greater flexibility for projects needing more material types while maintaining the current advantages.
Thank you very much for your time and consideration. I look forward to hearing your thoughts!

Best regards,
Ginmus

**Comment by splashdust** [comment]
Excellent and well put question!

1\) As you speculate in your question, the `u8` was chosen for memory conservation, since this data will increase memory usage linearly with the number of stored voxels (barring any other optimizations). The reason it returns `u32` is simply because the output is sent to the shader, and that makes 32-bit values fast and convenient. The returned value is not permanently stored in memory, so the size is less of an issue there.

2\) Currently I have no good "off the shelf" solution for you. Depending on your needs, [multiple worlds](https://github.com/splashdust/bevy_voxel_world/blob/main/examples/multiple_worlds.rs) may be a solution (then you could have 256 materials per world). You can also write a [custom shader](https://github.com/splashdust/bevy_voxel_world/blob/main/examples/custom_material.rs) and interpret the material data differently there, which may help, but the `u8` limitation is still there.

3, 4) I've got no concrete plans for the moment, but have been thinking about something along those lines. For example, enabling custom voxel data storage together with ways of sending that to the shader would make custom shaders much more powerful. I am open to suggestions and PRs! :)

**Comment by juzi5201314** [comment]
Hi @splashdust ,I have an idea to change the texture index type to a generic type, which will allow a larger number of material types without increasing the memory usage at a small number of material types (maybe?).
https://github.com/juzi5201314/bevy_voxel_world/commit/fd44af500c6d2f450c8cbfcb3a5cdede11ad2721

**Comment by splashdust** [comment]
> Hi @splashdust ,I have an idea to change the texture index type to a generic type, which will allow a larger number of material types without increasing the memory usage at a small number of material types (maybe?). [juzi5201314@fd44af5](https://github.com/juzi5201314/bevy_voxel_world/commit/fd44af500c6d2f450c8cbfcb3a5cdede11ad2721)

Finally had a chance to look at this, and it seems like a good idea! Would you mind opening a PR with this change?

**Comment by juzi5201314** [comment]
> > Hi @splashdust ,I have an idea to change the texture index type to a generic type, which will allow a larger number of material types without increasing the memory usage at a small number of material types (maybe?). [juzi5201314@fd44af5](https://github.com/juzi5201314/bevy_voxel_world/commit/fd44af500c6d2f450c8cbfcb3a5cdede11ad2721)
> 
> Finally had a chance to look at this, and it seems like a good idea! Would you mind opening a PR with this change?

👌#39

**Comment by splashdust** [comment]
Closing this as fixed in #39

Available in [0.9.0](https://github.com/splashdust/bevy_voxel_world/releases/tag/0.9.0)
